src/phoenix/trace/llama_index/callback.py

```diff
@@ -381,3 +410,43 @@ def _get_response_output(response: Any) -> Iterator[Tuple[str, Any]]:
     else:
         yield OUTPUT_VALUE, str(response)
         yield OUTPUT_MIME_TYPE, MimeType.TEXT
+
+
+def _get_end_time(event_data: CBEventData, span_events: List[SpanEvent]) -> Optional[datetime]:
+    """
+    A best-effort attempt to get the end time of an event.
+
+    LlamaIndex's callback system does not guarantee that the on_event_end hook is always called, for
+    example, when an error occurs mid-event.
+    """
+    if end_event := event_data.get("end_event"):
+        tz_naive_end_time = _timestamp_to_tz_naive_datetime(end_event.time)
+    elif span_events:
+        last_span_event = sorted(span_events, key=lambda event: event.timestamp)[-1]
+        tz_naive_end_time = last_span_event.timestamp
+    else:
+        return None
+    return _tz_naive_to_tz_aware_datetime(tz_naive_end_time)
+
+
+def _timestamp_to_tz_aware_datetime(timestamp: str) -> datetime:
+    """Converts a timestamp string to a timezone-aware datetime."""
+    return _tz_naive_to_tz_aware_datetime(_timestamp_to_tz_naive_datetime(timestamp))
+
+
+def _timestamp_to_tz_naive_datetime(timestamp: str) -> datetime:
+    """Converts a timestamp string to a timezone-naive datetime."""
+    return datetime.strptime(timestamp, TIMESTAMP_FORMAT)
+
+
+def _tz_naive_to_tz_aware_datetime(timestamp: datetime) -> datetime:
+    """Converts a timezone-naive datetime to a timezone-aware datetime."""
+    return timestamp.replace(tzinfo=_LOCAL_TZINFO)
+
+
+def _get_stacktrace(exception: BaseException) -> str:
```

Comment:
Not sure, let's use that one if it's recommend by OTEL.

Reply:
Looks like `format_exc` is just a wrapper around `format_exception`. `format_exception` is the one we want to use in this case since it takes in the exception object itself.

```python
def format_exc(limit=None, chain=True):
    """Like print_exc() but return a string."""
    return "".join(format_exception(*sys.exc_info(), limit=limit, chain=chain))
```

Reply:
```
traceback.format_exception(exc, /, [value, tb, ]limit=None, chain=True)
Format a stack trace and the exception information. The arguments have the same meaning as the corresponding arguments to [print_exception()](https://docs.python.org/3/library/traceback.html#traceback.print_exception). The return value is a list of strings, each ending in a newline and some containing internal newlines. When these lines are concatenated and printed, exactly the same text is printed as does [print_exception()](https://docs.python.org/3/library/traceback.html#traceback.print_exception).
```

The lines returned by `format_exception` contain newline characters, hence, joining by `""`. Will change back.